File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    // Ask the user for a card number
    long n;
    n = get_long("Card number to check: ");
    long n_test = n;

    // Check that the number of digits matches valid card numbers
    // American Express uses 15-digit numbers, MasterCard uses 16-digit numbers, and Visa uses 13- and 16-digit numbers
    int n_length = 0;
    while (n_test > 0)
    {
        n_test = (n_test / 10);
        n_length = n_length + 1;
    }
    if (n_length != 13 && n_length != 15 && n_length != 16)
    {
        printf("INVALID\n");
    }
    // If the number is valid, proceed to analyse its digits and compute checksum
    else
    {
        long checksum;
        long checksum_odd;
        long checksum_even;
        int i = 1, current_digit = 0;
        long n_test_2 = n;
        for (i = 1; i < n_length + 1; i++)
        {
            current_digit = n_test_2 % 10;
            n_test_2 = n_test_2 / 10;
            // Multiply every other digit by 2 starting with the second-to-last digit
            // Deal with these "even" digits, i.e. every other digit starting with the second-to-last digit
            if (i % 2 == 0)
            {
                int current_checksum_even;
                current_checksum_even = (current_digit * 2);
                // Add those products' digits together: do the sum of their digits, not of their values
                // Dealing first with 2-digit numbers: separate first and second digits, then add them together
                int current_checksum_even_length = 0;
                if (current_checksum_even > 9)
                {
                    current_checksum_even = (current_checksum_even / 10) + current_checksum_even % 10;
                    checksum_even = checksum_even + current_checksum_even;
                }
                // Then dealing with 1-digit numbers: do a simple sum
                else
                {
                    checksum_even = checksum_even + current_checksum_even;
                }
            }
            // Do the sum of "odd" digits, i.e. every other digit starting with the last digit, without multiplying them by 2
            else
            {
                checksum_odd = checksum_odd + current_digit;
            }
        }
        // Compute total checksum, adding partial checksums
        checksum = checksum_odd + checksum_even;

        // If the total's last digit is 0, the number is valid
        if (checksum % 10 == 0)
        {
            long first_digit = n;
            long second_digit = n;
            while (first_digit >= 10)
            {
                first_digit = first_digit / 10;
                printf("first_digit : %ld\n", first_digit);
            }
            while (second_digit >= 100)
            {
                second_digit = second_digit / 10;
                printf("second_digit : %ld\n", second_digit);
            }
            second_digit = second_digit % 10;
            printf("second_digit : %ld\n", second_digit);

            // If numbers start with 4, identify the card as a Visa
            if (first_digit == 4)
            {
                printf("VISA\n");
            }
            // If numbers start with 34 or 37, identify the card as an American Express
            else if (first_digit == 3 && (second_digit == 4 || second_digit == 7))
            {
                printf("AMEX\n");
            }
            // If numbers start with 51, 52, 53, 54, or 55, identify the card as a MasterCard
            else if (first_digit == 5 &&
                     (second_digit == 1 ||
                      second_digit == 2 ||
                      second_digit == 3 ||
                      second_digit == 4 ||
                      second_digit == 5))
            {
                printf("MASTERCARD\n");
            }
            // If the first digits are not the expected ones, declare the number invalid
            else
            {
                printf("INVALID\n");
            }
        }
        // If the total's last digit is not 0, declare the number invalid
        else
        {
            printf("INVALID\n");
        }
    }
}

File: sentimental-credit/credit.py
import cs50


def main():

    # Ask the user for a card number
    n = cs50.get_string("Card number to check: ")

    # Check that the number of digits matches valid card numbers
    # American Express uses 15-digit numbers, MasterCard uses 16-digit numbers, and Visa uses 13- and 16-digit numbers
    n_length = len(n)
    print("n_length : " + str(n_length))

    if (n_length != 13 and n_length != 15 and n_length != 16):
        print("INVALID")

    # If the number is valid, proceed to analyse its digits and compute checksum
    else:
        checksum = 0

        # Even card numbers
        if (n_length % 2 == 0):
            # For "even" digits, multiply every other digit by 2
            checksum_even_str = ""
            for i in reversed(range(0, n_length, 2)):
                current_digit = int(n[i])
                checksum_even_str = checksum_even_str + (str(current_digit * 2))
            # Add those products' digits together: do the sum of their digits, not of their values
            checksum_even = 0
            for j in checksum_even_str:
                checksum_even = checksum_even + int(j)

            # For "odd" digits, do a simple sum
            checksum_odd = 0
            for i in reversed(range(1, n_length, 2)):
                current_digit = int(n[i])
                checksum_odd = checksum_odd + current_digit

        # Odd card numbers
        else:
            # For "even" digits, multiply every other digit by 2
            checksum_even_str = ""
            for i in reversed(range(1, n_length, 2)):
                current_digit = int(n[i])
                checksum_even_str = checksum_even_str + (str(current_digit * 2))
            # Add those products' digits together: do the sum of their digits, not of their values
            checksum_even = 0
            for j in checksum_even_str:
                checksum_even = checksum_even + int(j)

            # For "odd" digits, do a simple sum
            checksum_odd = 0
            for i in reversed(range(0, n_length, 2)):
                current_digit = int(n[i])
                checksum_odd = checksum_odd + current_digit

        # Compute total checksum, adding partial checksums
        checksum = checksum_odd + checksum_even

        # Select digits of checksum which will be tested
        last_digit_checksum = int(str(checksum)[len(str(checksum))-1])
        first_digit_n = int(n[0])
        second_digit_n = int(n[1])

        # Check if the number is valid, i.e. if the total's last digit is 0
        if (last_digit_checksum == 0):

            # If a valid number starts with 4, identify the card as a Visa
            if (first_digit_n == 4):
                print("VISA")

            # If a valid number starts with 34 or 37, identify the card as an American Express
            elif (first_digit_n == 3 and (second_digit_n == 4 or second_digit_n == 7)):
                print("AMEX")

            # If a valid number starts with 51, 52, 53, 54, or 55, identify the card as a MasterCard
            elif (first_digit_n == 5 and (second_digit_n == 1 or second_digit_n == 2 or second_digit_n == 3 or second_digit_n == 4 or second_digit_n == 5)):
                print("MASTERCARD")

            # Declare the number invalid if the first digits of a valid number are not the ones expected for these card providers
            else:
                print("INVALID")

        # Declare the number invalid if the total's last digit is not 0
        else:
            print("INVALID")

    return 0


if __name__ == "__main__":
    main()